db_functions.py

```diff
@@ -0,0 +1,458 @@
+
+    conn = db_connect()
+
+    author = header["author"]
+    version = header["version"]
+
+    src = "dat" if author not in ["scan", "scummvm"] else author
+
+    detection = (src == "scummvm")
+    status = "detection" if detection else src
+
+    conn.cursor().execute(f"SET @fileset_time_last = {int(time.time())}")
+
+    with conn.cursor() as cursor:
+        cursor.execute("SELECT MAX(`transaction`) FROM transactions")
+        transaction_id = cursor.fetchone()['MAX(`transaction`)'] + 1
+
+    category_text = f"Uploaded from {src}"
+    log_text = f"Started loading DAT file, size {os.path.getsize(filepath)}, author '{author}', version {version}. State '{status}'. Transaction: {transaction_id}"
+
+    user = f'cli:{getpass.getuser()}'
+    create_log(escape_string(category_text), user, escape_string(log_text), conn)
+
+    for fileset in game_data:
+        if detection:
+            engine_name = fileset["engine"]
+            engineid = fileset["sourcefile"]
+            gameid = fileset["name"]
+            title = fileset["title"]
+            extra = fileset["extra"]
+            platform = fileset["platform"]
+            lang = fileset["language"]
+
+            insert_game(engine_name, engineid, title, gameid, extra, platform, lang, conn)
+        elif src == "dat":
+            if 'romof' in fileset and fileset['romof'] in resources:
+                fileset["rom"] = fileset["rom"] + resources[fileset["romof"]]["rom"]
+
+        key = calc_key(fileset) if detection else ""
+        megakey = calc_megakey(fileset['rom']) if not detection else ""
+        log_text = f"size {os.path.getsize(filepath)}, author '{author}', version {version}. State '{status}'."
+
+        if insert_fileset(src, detection, key, megakey, transaction_id, log_text, conn):
+            for file in fileset["rom"]:
+                insert_file(file, detection, src, conn)
+                for key, value in file.items():
+                    if key not in ["name", "size"]:
+                        insert_filechecksum(file, key, conn)
+
+    if detection:
+        conn.cursor().execute("UPDATE fileset SET status = 'obsolete' WHERE `timestamp` != FROM_UNIXTIME(@fileset_time_last) AND status = 'detection'")
+    cur = conn.cursor()
+    
+    try:
+        cur.execute(f"SELECT COUNT(fileset) from transactions WHERE `transaction` = {transaction_id}")
+        fileset_insertion_count = cur.fetchone()['COUNT(fileset)']
+        category_text = f"Uploaded from {src}"
+        log_text = f"Completed loading DAT file, filename '{filepath}', size {os.path.getsize(filepath)}, author '{author}', version {version}. State '{status}'. Number of filesets: {fileset_insertion_count}. Transaction: {transaction_id}"
+    except Exception as e:
+        print("Inserting failed:", e)
+    else:
+        user = f'cli:{getpass.getuser()}'
+        create_log(escape_string(category_text), user, escape_string(log_text), conn)
+
+def compare_filesets(id1, id2, conn):
```

Comment:
I think this will not scale well since you're basically doing quadratic computation.

I propose the following approach:
*Pick a file from the incoming fileset
* Select from the fileset set of fileset ids that have these files with checksum matching. should be a single SELECT
* If the list is too big, like, over 100 entries (low chance), add one more file
* Iterate over the resulting files, or even ask SQL to do it with UNION ALL

Eventually, we will have like 15k filesets, so quadratic computations are not feasible